### PR TITLE
Fix llvm relo error caused by multiple ports in policy

### PR DESCRIPTION
**Pull request by jrfastab**
This is a similar error to the one we already fixed for IPv6 addresses. Here we implement a similar fix. The macros in this case however are a bit more complex because the port policy is a variable length map vs a single structure. This forces the macro to handle variable length maps as well and results in a slightly complex macro.

An alternative solution would have been to extend the code generation to emit more of the bpf codes. This would be a more significant change in design and I think the macro is good enough for the time being.

### CODE REVIEW

**Comment by jrfastab** [comment]
interestingly 4.8 works, 4.9 gives the errors from jenkins, 4.11 and 4.12 work again.

**Comment by jrfastab** [comment]
OK got this working on 4.9 by "unrolling" the for loop over `allowed` in the macro to make it a series of ternary expressions. Will do some more testing but should resolve the issue reported here although it is not pretty.

File: pkg/endpoint/bpf.go [hunk 1]
```diff
@@ -63,19 +64,21 @@ func (e *Endpoint) writeL4Map(fw *bufio.Writer, owner Owner, m policy.L4PolicyMa
...
 		redirect = common.Swab16(redirect)
-		entry := fmt.Sprintf("{%d,%d,%d}", dport, redirect, protoNum)
+		entry := fmt.Sprintf("%d,%d,%d,%d", index, dport, redirect, protoNum)
 		if array != "" {
 			array = array + "," + entry
 		} else {
 			array = entry
 		}
 
+		index += 1
```
[the remark this answers is not in the document]
Comment (aanm) [reply pkg/endpoint/bpf.go]: @jrfastab If the intention is to use the index of the array you can replace the line 49 from `for _, l4 := range m {` to `for i, l4 := range m {`

**Review by tgraf — APPROVED** [review]
I agree with Andre's comment, otherwise this looks good to me.

File: pkg/endpoint/bpf.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (jrfastab) [reply pkg/endpoint/bpf.go]: Is `i` here guaranteed to iterate from 0,1,2...,n? I thought (with my limited go knowledge) that range was allowed to iterate through a map in any arbitrary order.

Also note I'll need to do another push with the new macros shortly.

[the remark this answers is not in the document]
Comment (tgraf) [reply pkg/endpoint/bpf.go]: I think @aanm and I both missed that this is a map and not slice. The current code is OK

[the remark this answers is not in the document]
Comment (aanm) [reply pkg/endpoint/bpf.go]: Oh it's a map, you're right then.

```
for i, element := range []string{"foo", "bar"}{
    // i is 0, 1..
    // element is "foo", "bar"..
}
for key, value := range map[string]string{"foo": "value1", "bar": "value2"}{
    // key is "foo", "bar"..
    // value is "value1", "value2"..
}
```

[the remark this answers is not in the document]
Comment (jrfastab) [reply pkg/endpoint/bpf.go]: OK great I guess I can make it i++ per the suggestion from the bot. Although it doesn't much matter to me.

**Pull request merged**